### PR TITLE
Implement a login device flow for the go SDK

### PR DESCRIPTION
**Pull request by loganintech**

### CODE REVIEW

File: login.go [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+}
+
+func getDeviceCode(ctx context.Context, client *ConductoroneAPI, clientID string) (*DeviceCodeResponse, error) {
+	httpClient := client.sdkConfiguration.DefaultClient
+
+	deviceCodeURL := "https://" + client.sdkConfiguration.ServerURL + "/auth/v1/device_authorization"
+	vals := url.Values{}
+	vals.Add("client_id", clientID)
+
+	req, err := http.NewRequestWithContext(ctx, "POST", deviceCodeURL, strings.NewReader(vals.Encode()))
```
Comment (pquerna) [inline login.go:67]: do we have a common constructor for requests that should apply headers like user-agent?

File: login.go [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+
+		return tokenResp, nil
+	}
+}
+
+func doClientCredentialRequest(ctx context.Context, client *ConductoroneAPI, tokenResp *tokenResponse) (*clientResp, error) {
+	httpClient := client.sdkConfiguration.DefaultClient
+	pccURL := "https://" + client.sdkConfiguration.ServerURL + "/api/v1/iam/personal_clients"
+
+	personalClientReq, err := http.NewRequestWithContext(ctx, "POST", pccURL, bytes.NewReader([]byte("{\"display_name\": \"Created From Cone\"}")))
```
Comment (pquerna) [inline login.go:155]: this string should probably be a parameter to the Login func?

Comment (loganintech) [reply login.go:155]: Yeah good call

File: login.go [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+	UserCode        string `json:"user_code"`
+	VerificationURI string `json:"verification_uri_complete"`
+	ExpiresIn       int64  `json:"expires_in"`
+	Interval        int64  `json:"interval"`
+}
+
+func LoginFlow(ctx context.Context, tenantName string, clientID string) (*ClientCredentials, error) {
+	opts := []SDKOption{}
+	// If they pass a URL, use the whole URL
+	if strings.Contains(tenantName, ".") {
```
[the remark this answers is not in the document]
Comment (pquerna) [reply login.go:32]: i think we could make a NormalizeTenantParam (or similar name), which takes a tenant name, tenant full domain, tenant url as a url, etc

[the remark this answers is not in the document]
Comment (pquerna) [reply login.go:32]: We could even make it a WithTenant() option function, which normalized and  converted itself into one of the other With** functions :)

File: login.go [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+		return nil, err
+	}
+
+	codeResp := &DeviceCodeResponse{}
+	err = json.Unmarshal(data, codeResp)
+	if err != nil {
+		return nil, err
+	}
+
+	fmt.Printf("Open: %s\n", codeResp.VerificationURI)
```
Comment (pquerna) [inline login.go:98]: should we use webbrowser here?

i think at the very least, Login() should take a callback function that is OpenURL(string)?  then the caller can control the UX?

Comment (pquerna) [reply login.go:98]: webborwser :https://pkg.go.dev/github.com/toqueteos/webbrowser

Comment (loganintech) [reply login.go:98]: I like the callback idea over the web browser idea by default. Although we can use webbrowser in cone.

File: login.go [hunk 5]
```diff
@@ -26,7 +26,7 @@ type DeviceCodeResponse struct {
 	Interval        int64  `json:"interval"`
 }
 
-func LoginFlow(ctx context.Context, tenantName string, clientID string, personalClientCredentialDisplayName string) (*ClientCredentials, error) {
+func LoginFlow(ctx context.Context, tenantName string, clientID string, personalClientCredentialDisplayName string, cb func(authorizeUrl string) error) (*ClientCredentials, error) {
```
Comment (pquerna) [inline login.go:29]: `authorizeUrl` -> `validateUrl`

File: extra_sdk_options.go [hunk 6]
```diff
@@ -0,0 +1,34 @@
+package conductoroneapi
+
+import (
+	"net/url"
+	"strings"
+)
+
+const GoSDKClientID = "2RCzHlak5q7CY14SdBc8HoZEJRf"
```
Comment (pquerna) [inline extra_sdk_options.go:8]: weird exported const name.  maybe lets start without exporting it?

at some level would expect it to be ClientIdGolangSDK? but dunno.

File: login.go [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+
+	startLoop := time.Now()
+	for {
+		select {
+		case <-time.After(time.Duration(deviceCodeResp.Interval) * time.Second):
+			if time.Now().After(startLoop.Add(time.Duration(deviceCodeResp.ExpiresIn) * time.Second)) {
+				return nil, errors.New("timeout")
+			}
+		case <-ctx.Done():
+			return nil, errors.New("context canceled")
```
Comment (pquerna) [inline login.go:130]: return nil, ctx.Err()

File: login.go [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+func doClientCredentialRequest(ctx context.Context, client *ConductoroneAPI, tokenResp *tokenResponse, personalClientCredentialDisplayName string) (*clientResp, error) {
+	httpClient := client.sdkConfiguration.DefaultClient
+	baseURL := utils.ReplaceParameters(client.sdkConfiguration.GetServerDetails())
+
+	pccURL, err := url.JoinPath(baseURL, "/api/v1/iam/personal_clients")
+	if err != nil {
+		return nil, err
+	}
+
+	personalClientReq, err := newReq(ctx, "POST", pccURL, bytes.NewReader([]byte(fmt.Sprintf("{\"display_name\": \"%s\"}", personalClientCredentialDisplayName))))
```
Comment (pquerna) [inline login.go:180]: nit: this won't be property escaped.

should just make an local struct with json tags?

File: extra_sdk_options.go (hunk 6 above)
Comment (loganintech) [reply extra_sdk_options.go:8]: It's already pretty hard to use our SDK considering they need to do all the funkyness with token source, wouldn't removing this export just make it more difficult?

File: login_test.go [hunk 9]
```diff
@@ -0,0 +1,25 @@
+package conductoroneapi
+
+import (
+	"context"
+	"fmt"
+	"testing"
+)
+
+func TestLoginFlow(t *testing.T) {
```
Comment (pquerna) [inline login_test.go:9]: feels like an odd test case to commit?

it'll run, create dead end devicecodes, but... never finish them?

and... it'll timeout?

feels like the test should be Skip() and return early for now?

**Review by pquerna — APPROVED** [review]

File: login_test.go (hunk 9 above)
Comment (loganintech) [reply login_test.go:9]: 'twas just a way for me to quickly test the SDK side of this work. Happy to add a skip call in to keep for posterity

**Pull request merged**